File: src/main/java/com/multicampus/gamesungcoding/a11ymarketserver/feature/seller/entity/Seller.java
package com.multicampus.gamesungcoding.a11ymarketserver.feature.seller.entity;

import com.multicampus.gamesungcoding.a11ymarketserver.common.id.UuidV7;
import com.multicampus.gamesungcoding.a11ymarketserver.feature.product.entity.Product;
import com.multicampus.gamesungcoding.a11ymarketserver.feature.user.entity.UserRole;
import com.multicampus.gamesungcoding.a11ymarketserver.feature.user.entity.Users;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "sellers")
@EntityListeners(AuditingEntityListener.class)
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Seller {

    @Id
    @UuidV7
    @Column(nullable = false, updatable = false, length = 16)
    private UUID sellerId;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    @OnDelete(action = OnDeleteAction.CASCADE)
    private Users user;

    @Column(nullable = false)
    private String sellerName;

    @Column(nullable = false)
    private String businessNumber;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    private SellerGrades sellerGrade;

    @Column(length = 1024)
    private String sellerIntro;

    @Column(nullable = false, name = "is_a11y_guarantee")
    private Boolean isA11yGuarantee;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private SellerSubmitStatus sellerSubmitStatus;

    @CreatedDate
    @Column(updatable = false)
    private LocalDateTime submitDate;

    private LocalDateTime approvedDate;

    @LastModifiedDate
    private LocalDateTime updatedAt;

    @OneToMany(mappedBy = "seller")
    private List<Product> products;

    @Builder
    private Seller(Users user,
                   String sellerName,
                   String businessNumber,
                   String sellerIntro) {

        this.user = user;
        this.sellerName = sellerName;
        this.businessNumber = businessNumber;
        this.sellerGrade = SellerGrades.NEWER;
        this.sellerIntro = sellerIntro;
        this.isA11yGuarantee = false;
        this.sellerSubmitStatus = SellerSubmitStatus.PENDING;
    }

    public void approve() {
        this.sellerSubmitStatus = SellerSubmitStatus.APPROVED;
        this.approvedDate = LocalDateTime.now();
        this.user.changeRole(UserRole.SELLER);
    }

    public void reject() {
        this.sellerSubmitStatus = SellerSubmitStatus.REJECTED;
        this.approvedDate = LocalDateTime.now();
    }

    public void updateAdminSellerInfo(String sellerName,
                                      String businessNumber,
                                      String sellerIntro,
                                      SellerGrades sellerGrade,
                                      Boolean a11yGuarantee) {

        if (sellerName != null) this.sellerName = sellerName;
        if (businessNumber != null) this.businessNumber = businessNumber;
        if (sellerIntro != null) this.sellerIntro = sellerIntro;
        if (sellerGrade != null) this.sellerGrade = sellerGrade;
        if (a11yGuarantee != null) this.isA11yGuarantee = a11yGuarantee;
    }
}

File: src/main/kotlin/com/multicampus/gamesungcoding/a11ymarketserver/feature/seller/entity/Seller.kt
package com.multicampus.gamesungcoding.a11ymarketserver.feature.seller.entity

import com.multicampus.gamesungcoding.a11ymarketserver.common.id.UuidV7
import com.multicampus.gamesungcoding.a11ymarketserver.feature.product.entity.Product
import com.multicampus.gamesungcoding.a11ymarketserver.feature.user.entity.UserRole
import com.multicampus.gamesungcoding.a11ymarketserver.feature.user.entity.Users
import jakarta.persistence.*
import org.hibernate.annotations.OnDelete
import org.hibernate.annotations.OnDeleteAction
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.annotation.LastModifiedDate
import org.springframework.data.jpa.domain.support.AuditingEntityListener
import java.time.LocalDateTime
import java.util.*

@Entity
@Table(name = "sellers")
@EntityListeners(AuditingEntityListener::class)
class Seller(
    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    @OnDelete(action = OnDeleteAction.CASCADE)
    val user: Users,

    @Column(nullable = false)
    var sellerName: String,

    @Column(nullable = false)
    var businessNumber: String,

    @Column(length = 1024)
    var sellerIntro: String? = null
) {
    @Id
    @UuidV7
    @Column(nullable = false, updatable = false, length = 16)
    val sellerId: UUID? = null

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    var sellerGrade: SellerGrades = SellerGrades.NEWER
        private set

    @Column(nullable = false, name = "is_a11y_guarantee")
    var isA11yGuarantee: Boolean = false
        private set

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    var sellerSubmitStatus: SellerSubmitStatus = SellerSubmitStatus.PENDING
        private set

    @CreatedDate
    @Column(updatable = false)
    var submitDate: LocalDateTime? = null
        private set

    var approvedDate: LocalDateTime? = null
        private set

    @LastModifiedDate
    var updatedAt: LocalDateTime? = null
        private set

    @OneToMany(mappedBy = "seller")
    var products: MutableList<Product> = mutableListOf()
        private set

    // 비즈니스 로직
    fun approve() {
        this.sellerSubmitStatus = SellerSubmitStatus.APPROVED
        this.approvedDate = LocalDateTime.now()
        this.user.changeRole(UserRole.SELLER)
    }

    fun reject() {
        this.sellerSubmitStatus = SellerSubmitStatus.REJECTED
        this.approvedDate = LocalDateTime.now()
    }

    fun updateAdminSellerInfo(
        sellerName: String?,
        businessNumber: String?,
        sellerIntro: String?,
        sellerGrade: SellerGrades?,
        a11yGuarantee: Boolean?
    ) {
        sellerName?.let { this.sellerName = it }
        businessNumber?.let { this.businessNumber = it }
        sellerIntro?.let { this.sellerIntro = it }
        sellerGrade?.let { this.sellerGrade = it }
        a11yGuarantee?.let { this.isA11yGuarantee = it }
    }

    fun updateSellerInfo(sellerName: String?, sellerIntro: String?) {
        sellerName?.let { this.sellerName = it }
        sellerIntro?.let { this.sellerIntro = it }
    }
}
